sql/core/src/main/scala/org/apache/spark/sql/Melt.scala

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion, UnresolvedAttribute}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Cast, Expression, Literal, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.catalyst.plans.logical.Expand
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, StringType}
+
+private[sql] object Melt {
+  def of[_](
+      ds: Dataset[_],
+      ids: Seq[String],
+      values: Seq[String],
+      dropNulls: Boolean,
+      variableColumnName: String,
+      valueColumnName: String): DataFrame = {
+    // values should be disjoint to ids
+    if (values.intersect(ids).nonEmpty) {
+      throw new AnalysisException("MELT_ID_AND_VALUE_COLUMNS_NOT_DISJOINT", Array(
+        ids.mkString(", "), values.mkString(", "), values.intersect(ids).mkString(", ")
+      ))
+    }
+
+    // id columns must not contain variable or value column name
+    if (ids.contains(variableColumnName)) {
+      throw new AnalysisException("MELT_VARIABLE_COLUMN_IS_ID_COLUMN", Array(
+        variableColumnName, ids.mkString(", ")
+      ))
+    }
+    if (ids.contains(valueColumnName)) {
+      throw new AnalysisException("MELT_VALUE_COLUMN_IS_ID_COLUMN", Array(
+        valueColumnName, ids.mkString(", ")
+      ))
+    }
+
+    // if no values given, all non-id columns are melted
+    val valueNameCandidates = ds.columns.diff(ids).toSeq
+    val valueNames = Some(values).filter(_.nonEmpty).getOrElse(valueNameCandidates)
+
+    // if there are no values given and no non-id columns exist, we cannot melt
+    if (valueNames.isEmpty) {
+      throw new AnalysisException("MELT_REQUIRES_VALUE_COLUMNS", Array(ids.mkString(", ")))
+    }
+
+    // resolve given id and value column names
+    val idExprs = ids.map(resolve(ds, _))
+    val valueExprs = valueNames.map(resolve(ds, _))
+
+    // all given ids and values should exist in ds
+    val unresolvedIds = idExprs.filter(_.isLeft).map(_.swap.toOption.get)
+    if (unresolvedIds.nonEmpty) {
+      val idNameCandidates = ds.columns.diff(valueNames).toSeq
+      throw new AnalysisException("MISSING_COLUMNS",
+        Array(unresolvedIds.mkString(", "), idNameCandidates.mkString(", ")))
+    }
+    val unresolvedValues = valueExprs.filter(_.isLeft).map(_.swap.toOption.get)
+    if (unresolvedValues.nonEmpty) {
+      throw new AnalysisException("MISSING_COLUMNS",
+        Array(unresolvedValues.mkString(", "), valueNameCandidates.mkString(", ")))
+    }
+
+    // all melted values have to have the same type
+    val resolvedValues = valueExprs.filter(_.isRight).map(_.toOption.get)
+    val valueType = resolvedValues.map(_.dataType).reduce(tightestCommonType)
```

Comment:
Spark SQL has a type coercion framework to do this kind of work, see `TypeCoercion.scala`

Reply:
And we can leverage this framework if melt is a logical plan

Reply:
The best I could do is use `findWiderType` / `findWiderTypeWithoutStringPromotion`, applied to `Melt` value columns in a new type coercion rule `MeltCoercion`.